Defer local updates if there are missing updates and only call `GetMissingOnyxMessages` once

chrispader opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@danieldoglas 

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/38748
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

#### Test 1

I created a [testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing) that prevents some updates from being applied (therefore `GetMissingOnyxUpdates` will be triggered) and has some useful logs.

Either test in this branch or on your own. In [`OnyxUpdateManager`](https://github.com/Expensify/App/blob/main/src/libs/actions/OnyxUpdateManager.ts) do the following:

1. Delay the call to `applyDeferredUpdates` with a `setTimeout`
2. Prevent some update from being applied
3. Check that `GetMissingOnyxUpdates` is triggered
4. Add some more updates
5. Check that these updates are added to the `deferredUpdates` object
6. Check that no more calls to `GetMissingOnyxUpdates` are performed
7. Once the timeout has ended, check that both the missing and deferred updates are applied correctly.


#### Test 2
1. Create a chat report with a peer
2. Create a policy room between both of you
3. after those rooms are created, and both of you opened them at least once, comment these calls to make sure we're not reloading the data from the backend instead: https://github.com/Expensify/App/blob/a4d01a87b242eef06af37db965bd135745e342e5/src/libs/actions/Report.ts#L769-L789
4. Now send messages in both chat rooms, alternating between them, and fast, in a way that you can notice if anything is missing (so both start sending some kind of ordered messages, like sequential numbers)
5. Make sure that none of the messages were lost in those chats.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Not needed.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
8. Upload an image via copy paste
9. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

Only Test 2 from "Tests"

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

<img width="557" alt="Screenshot 2024-03-27 at 16 05 13" src="https://github.com/Expensify/App/assets/20173411/56863211-c81f-454e-aa4c-5a24cf298aba">


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

![Simulator Screenshot - iPhone 15 Pro - 2024-03-27 at 16 10 34](https://github.com/Expensify/App/assets/20173411/a5580be6-c9f9-4ffe-a845-7f1741261474)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

<img width="1164" alt="Screenshot 2024-03-27 at 16 13 55" src="https://github.com/Expensify/App/assets/20173411/a744d1fa-1e0c-4604-b7fe-d8eeb8620a52">

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

Tested deferring logic on web, iOS and Android. mWeb behavior is the same as in web...

Commit:
defer updates if missing updates are fetched

Commit:
simplify

Commit:
move deferred updates out of callback

Commit:
simplify comments

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

Commit:
fix: apply deferred updates in order

chrispader commented:
@danieldoglas could you take a look at the current implementation and just check, if this is at least somehow what you expected? just checking before i spend too much time on the actual implementation.

I expect this PR to be done tomorrow if the current approach is feasible.

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +70,101 @@ export default () => {
+                if (deferredUpdateBeforeReconnect) {
+                    OnyxUpdates.apply(deferredUpdateBeforeReconnect).finally(unpauseQueueAndReset);
+                    deferredUpdateBeforeReconnect = undefined;
+                    return;
```
You should not return here, since `deferredUpdates` could still exist and would not be applied in this case

chrispader replied:
I think i might not understand case (1) correctly. Why would we not receive the `lastUpdateID` from Onyx updates in the first place? Isn't this something thats always provided with updates?

Also, what does the `ReconnectApp` query achieve in case (1)?

danieldoglas replied:
can you point me to what you're talking about when you say `Case (1)`?

chrispader replied:
oh sorry, i meant what's described here: https://github.com/Expensify/App/blob/9a1f43c286c599b0a694fad3ad5bd9ef7c0b309a/src/libs/actions/OnyxUpdateManager.ts#L57-L67

danieldoglas replied:
Ohh I see.

**Reliable** OnyxUpdate is something that is not enabled for everyone. So some users will log in and never receive previousUpdateID and lastUpdateID.  When we do enable it for them, once they receive the first request with lastUpdateID, we need to do a full reconnect to confirm that their client is in the right state before we start applying the updates.

danieldoglas replied:
What I meant with `You should not return here, since deferredUpdates could still exist and would not be applied in this case` is that while you're executing the full reconnect, it's possible that you're still gonna receive updates while that is running (since that's a slow GET). So we should still queue them up as we do for the getMissingOnyxMessages

chrispader replied:
Got it thanks, for the explanation. I already applied the change, so i'm not early-returning in this case..

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +70,101 @@ export default () => {
+                // Get the number of deferred updates before adding the new one
+                const existingDeferredUpdatesCount = deferredUpdates.length;
+                // Add the new update to the deferred updates
+                deferredUpdates.push(updateParams);
```
I think it's better if you have this as a keyValue object instead. If you do that, we won't need to sort this array later. Something like 
```suggestion
                defferedUpdates[updateParams.lastUpdateID] = updateParams;
```

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +70,101 @@ export default () => {
+                    return;
+                }
+
+                // If case (2) was triggered, then we need to sort the deferred updates by lastUpdateID and apply them in order.
+                const sortedDeferredUpdates = deferredUpdates.sort((a, b) => {
+                    const aLastUpdateID = Number(a.lastUpdateID) || 0;
+                    const bLastUpdateID = Number(b.lastUpdateID) || 0;
+
+                    return aLastUpdateID - bLastUpdateID;
                 });
```
```suggestion
```

If we store this as a keyValue object, we don't need to order it

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +70,101 @@ export default () => {
+                // In order for the deferred updates to be applied correctly in order,
+                // we need to check if there are any gaps deferred updates.
+                function validateUpdatesAreChained(): boolean {
+                    for (let i = 0; i < sortedDeferredUpdates.length - 1; i++) {
+                        const reversedDeferredUpdates = sortedDeferredUpdates.toReversed();
+                        const currentUpdate = reversedDeferredUpdates[i];
+                        const previousUpdate = reversedDeferredUpdates[i + 1];
+
+                        if (Number(previousUpdate.lastUpdateID) !== Number(currentUpdate.previousUpdateID)) {
+                            return false;
+                        }
+                    }
```
This should work since the keys will be ordered: 
![image](https://github.com/Expensify/App/assets/5818999/94f2b150-3ebb-4f18-a00a-b2d2b4626814)


```suggestion
                    let isFirst = true;
                    let previousValidLastUpdateID = 0;
                    for (const [key, value] of Object.entries(deferredUpdates)) {
                        // If it's the first one, we need to skip it since we won't have the previousUpdateID
                        // if the previousUpdateID key exists, then we can just move on since there's no gap
                        if (isFirst || deferredUpdates[value.previousUpdateID]) {
                            previousValidLastUpdateID = value.lastUpdateID;
                            isFirst = false;
                            continue;
                        }
                        return false;
                    }
```

danieldoglas reviewed (commented):
Added a few comments, but I think we're missing a few things:

- We should only apply the deferred updates if the current applied update is smaller than their lastUpdateID. Another way to see this is that we should only apply them if they are > the tha lastUpdateID in the `deferredUpdateBeforeReconnect`
- By applying the item above, we need to confirm that the updates that were received are also linked correctly to the `deferredUpdateBeforeReconnect`
- I'm not sure if there are gains in persisting the deferredUpdates or not. I think maybe not?

Commit:
simplify and improve applying deferred updates

Commit:
check if deferred updates are older than the currently applied one

chrispader commented:
> * We should only apply the deferred updates if the current applied update is smaller than their lastUpdateID. Another way to see this is that we should only apply them if they are > the tha lastUpdateID in the `deferredUpdateBeforeReconnect`

I updated the code so that we check for "old" deferred updates, that have already been applied through `GetMissingOnyxUpdates`. I basically just filter them out and store the rest in `pendingDeferredUpdates`, since they can only be outdated, if they were included in the response from `GetMissingOnyxUpdates`.

> * By applying the item above, we need to confirm that the updates that were received are also linked correctly to the `deferredUpdateBeforeReconnect`

Are case (1) and (2) expected to happen subsequently? As in my other comment, i might not fully understand case (1).

> * I'm not sure if there are gains in persisting the deferredUpdates or not. I think maybe not?

I don't think we need to persist the deferred values, since on a reload/restart we're fetching everything again anyway, right?

Updated the code and tested it thoroughly. I'm manually skipping some updates locally to simulate XHR requests not reaching the client. The missing updates are received from `GetMissingOnyxUpdates` and the (filtered) deferred updates are linked and applied correctly.

Commit:
improve code

chrispader commented:
> No need for C+ on this one!

Not sure if we need a C+ on this one? @danieldoglas 

https://github.com/Expensify/App/issues/38748#issuecomment-2012522080

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

danieldoglas commented:
I don't think we need a C+ on this one! Removing @getusha

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +70,114 @@ export default () => {
+
+                //  If we detect a gap in the deferred updates, re-fetch the missing updates.
+                if (!areDeferredUpdatesChained()) {
+                    canUnpauseQueuePromise = App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, lastValidDeferredUpdateID).finally(applyDeferredUpdates);
```
```suggestion
                    canUnpauseQueuePromise = App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, lastValidDeferredUpdateID);
```

hmmm this is not exactly what we need to do. If you do getMissingOnyxUpdates, it will apply the updates it will receive directly to onyx.

So let's say:

- Your current state is lastUpdateID 10
- You receive through pusher update A with previousUpdateID 15, lastUpdateID 18
- You then store that update and start a Request 1, getMissingOnyxMessages, from 10 to 15
- While you're executing Request 1, you receive update B with previousUpdateID 20 and lastUpdateID 22, which is stored in the deferredUpdates
- When Request 1 finishes, updates from 10 to 15 will be applied directly into onyx, and the local lastUpdateID is now 15
- You'll run this code, and see that there's a gap between update A (lastUpdateID 18) and update B (previousUpdateID 20)
- Because that gap exists, you'll do another getMissingOnyxUpdates, Request 2, from 18 to 20
- When request 2 finishes, it will return previousUpdateID 18 and lastUpdateID 20, and try to automatically apply that updates received through HTTPs before applying anything from the deferred queue. It will then notice that the local lastUpdateID is 15, which doesn't match the received previousUpdateID 18. It will query another GetMissingOnyxMessages from 15 to 18 (but this already exists in the deferredUpdates

So this will still trigger more getMissingOnyxUpdates, even if we have those updates ready to be applied.

chrispader replied:
You're right. I updated the logic. Can you give this another review? :)

Commit:
fix: refetching logic

chrispader commented:
@danieldoglas just triggered another review request. Could you give this a look, so we can maybe merge this today?

I'll try to be available in the evening, in case there are any issues.

Commit:
fix: comments and logic

Commit:
restructure file

Commit:
fix: re-write to detect possible multiple gaps in the deferred updates

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+    let firstUpdateAfterGaps: number | undefined;
+
+    for (const [index, update] of updateValues.entries()) {
+        // If it's the first update, we need to check that the previousUpdateID of the fetched update is the same as the lastUpdateIDAppliedToClient
+        // If any update's previousUpdateID doesn't match the lastUpdateID from the previous update, there's a gap in the deferred updates
+        const isFirst = index === 0;
+        if (isFirst && update.previousUpdateID !== lastUpdateIDAppliedToClient) {
+            break;
+        }
```
Does this means that if we have a gap in the first update already, we will stop here and not check if there are bigger gaps?

chrispader replied:
ah yes, you're right. this is wrong

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+            break;
+        }
+
+        if (updates[Number(update.previousUpdateID)]) {
+            // If a gap exists, we will not add any more updates to the applicable updates
+            // Instead, we we have to keep track of the first update after all detected gaps
+            if (gapExists) {
+                // If "firstUpdateAfterGaps" hasn't been set yet and there was a gap,
+                // we need to set it to the first update after all gaps
+                if (!firstUpdateAfterGaps) {
+                    firstUpdateAfterGaps = Number(update.previousUpdateID);
+                }
+            } else {
+                // If no gap exists yet and the previousUpdateID of the current update is found in the deferred updates, we can add it to the applicable updates
+                applicableUpdates[Number(update.lastUpdateID)] = update;
+            }
+        } else {
+            // When we find a(nother) gap, we need to set the flag to true and reset the "firstUpdateAfterGaps" variable, so that we can continue searching for the next update after all gaps
+            gapExists = true;
+            firstUpdateAfterGaps = undefined;
+        }
```
It took me a while to understand what is happening here, can you please add more comments explaining?

chrispader replied:
i simplified the conditions a bit, which removes unnecessary `if else` blocks. The reason we don't have early-returns here, is because `continue` is forbidden. This would make it more clean, but i think this is almost perfect

Commit:
simplified conditions

Commit:
further improve comments

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+        } else {
+            // When we find a (new) gap, we need to set "gapExists" to true and reset the "firstUpdateAfterGaps" variable,
+            // so that we can continue searching for the next update after all gaps
+            gapExists = true;
+            firstUpdateAfterGaps = undefined;
```
If we only have one update instead of N, we'll never fill the firstUpdateAfterGaps property, and will never add it to updatesAfterGaps, so we'll miss that update. I think we need to set the updateID here if isFirst is true.

danieldoglas replied:
If that happens we'll never fill the `latestMissingUpdateID` property too.

chrispader replied:
fixed!

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+            applyUpdates(applicableUpdates).then(() => {
+                // Same as above, we can ignore this case because
+                // it should not be possible for lastUpdateIDAppliedToClient to be null.
+                if (!lastUpdateIDAppliedToClient) {
```
also confirm if `latestMissingUpdateID > lastUpdateIDAppliedToClient`. Maybe parallel runs of other requests (like ReconnectApp) might have updated this variable

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+                // Same as above, we can ignore this case because
+                // it should not be possible for lastUpdateIDAppliedToClient to be null.
+                if (!lastUpdateIDAppliedToClient) {
+                    return;
```
also should we add `resolve()` here?

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+                // Then we can fetch the missing updates and apply them
+                App.getMissingOnyxUpdates(lastUpdateIDAppliedToClient, latestMissingUpdateID)
+                    .then(validateAndApplyDeferredUpdates)
+                    .then(() => resolve())
```
```suggestion
                    .then(resolve)
```

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,132 @@ Onyx.connect({
+    }
+
+    // If there are no gaps in the deferred updates, we can apply all deferred updates in order
+    return applyUpdates(applicableUpdates).then(() => undefined);
```
why are we doing this `.then(() => undefined)`?

chrispader replied:
Anyway, i changed the type of `queryPromise` to reflect the actual return values

danieldoglas commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +193,50 @@ export default () => {
+
                 // Since this is a full reconnectApp, we'll not apply the updates we received - those will come in the reconnect app request.
-                canUnpauseQueuePromise = App.finalReconnectAppAfterActivatingReliableUpdates();
+                queryPromise = App.finalReconnectAppAfterActivatingReliableUpdates().then(() => {
+                    // If "updateAfterReconnectApp" is set, it means that case (1) was triggered and we need to apply the update that was queued before the reconnectApp
+                    if (!deferredUpdateBeforeReconnect) {
+                        return;
+                    }
+
+                    OnyxUpdates.apply(deferredUpdateBeforeReconnect);
+                    deferredUpdateBeforeReconnect = undefined;
+                });
```
I've been thinking about this one specifically... I don't think we need to apply the deferred updates here. If you do a finalReconnect, you'll receive everything, including the update we've been storing, which is different than fetching the missing onyxUpdates that will only return the ones you're missing.

That being the case, let's drop the `deferredUpdateBeforeReconnect` property and remove this code after the finalReconnectApp.

danieldoglas reviewed (commented):
I think we're pretty close. 

I'm just concerned about 1 scenario I've just thought about. We're not currently checking if we have a `queryPromise` being executed.

So let's say this is the first time you're receiving a reliable onyxUpdate with lastUpdateID. We'll then schedule a reconnectApp (that takes a while to run). If you keep receiving pusher updates, for reconnectApp specifically, we'll continue to run new parallels reconnect app requests.

I think we need to set `queryPromise` to be a file variable and check if that variable is set before doing anything else. What do you think?

Commit:
fix: minor problems

Commit:
remove unnecessary code

Commit:
improve comment

Commit:
fix: check for queryPromise

chrispader commented:
@danieldoglas addressed all your comments and updated the code!

Commit:
update comment

Commit:
update comment

Commit:
move lines

danieldoglas commented:
Code is looking good to me, we need to test it before we merge it. Checking who's available in the C+ team

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

danieldoglas reviewed (commented):
Logic is looking good to me, let's have another set of eyes with the C+ team and see if this passes the tests before we move forward

mountiny commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+const applyUpdates = (updates: DeferredUpdatesDictionary) => Promise.all(Object.values(updates).map((update) => OnyxUpdates.apply(update)));
+
+// In order for the deferred updates to be applied correctly in order,
+// we need to check if there are any gaps deferred updates.
```
?
```suggestion
// we need to check if there are any gaps between deferred updates.
```

mountiny commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+    );
+
+    // If there are no remaining deferred updates after filtering out outdated ones,
+    //  we can just unpause the queue and return
```
```suggestion
    // we can just unpause the queue and return
```

mountiny commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -66,32 +200,39 @@ export default () => {
+
+                // Get the number of deferred updates before adding the new one
+                const existingDeferredUpdatesCount = Object.keys(deferredUpdates).length;
+                // Add the new update to the deferred updates
```
```suggestion

                // Add the new update to the deferred updates
```

Commit:
fix: comments

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

mountiny commented:
Been asked to check the code and comments. I think they are well written and I could follow what is going on, that being said, I think some visual chart representation in some README would be very helpful to convey the message across to the developer trying to understand what is going on here/ why we doing this

danieldoglas commented:
thanks @mountiny! we'll write those when we're working on the automated tests.

Commit:
fix: wrong early return

chrispader commented:
> I'm just concerned about 1 scenario I've just thought about. We're not currently checking if we have a `queryPromise` being executed.
> 
> I think we need to set `queryPromise` to be a file variable and check if that variable is set before doing anything else. What do you think?

@danieldoglas just noticed a mistake i made, where i would early-return before we could event add the deferred updates to the list. I changed this, so it only early-returns for `ReconnectApp`

Commit:
fix: missing return

Commit:
Revert "fix: missing return"

This reverts commit 252f0ab447048893ba3b2d86f778e6b391483463.

hungvu193 commented:
While testing this PR with @eh2077 , we're seeing this issue:

https://github.com/Expensify/App/assets/16502320/97821615-06c1-411a-a87e-7e1ced22980e

The chat order is correct, however screen is flashing and sometimes it jumps to the top.

eh2077 commented:
Yeah, this is a recording from my end. The screen is flashing

https://github.com/Expensify/App/assets/117511920/cd839ab8-16a9-4493-9d97-b200fd20d7a9

hungvu193 commented:
@chrispader  One question, can we still use your [testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing) to test `Test 1` after your last commit?

chrispader commented:
> @chrispader One question, can we still use your [testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing) to test `Test 1` after your last commit?

yes, i updated it accordingly. But let me clean it up a bit. There are a lot of logs i'm using right now to test my changes

chrispader commented:
> While testing this PR with @eh2077 , we're seeing this issue:
> 
>  Issue.2.-.Test.2.mov 
> 
> The chat order is correct, however screen is flashing and sometimes it jumps to the top.

Are you testing this in this PR directly or on the testing branch?

If you did in the testing branch, this is wrong. There are some manual `setTimeouts` (delays) and some messages are skipped, in order to test this manually.

If you actually want to QA-test the UI, please do that in this PR directly

eh2077 commented:
@chrispader I tested on this branch.

chrispader commented:
Just tested this on `main`, and i'm seeing these screen flashes and lags there too. So this seems unrelated to this PR.

On the left side, you can see `main` running on the local dev server. On the right is just `new.expensify.com`:

https://github.com/Expensify/App/assets/20173411/a45e667f-0a7d-4a64-94f4-5a7f15ce8204

hungvu193 commented:
Ok cool. Let's me check. Also can I use your test branch now?

hungvu193 commented:
> Just tested this on `main`, and i'm seeing these screen flashes and lags there too. So this seems unrelated to this PR.
> 
> 
> 
> On the left side, you can see `main` running on the local dev server. On the right is just `new.expensify.com`:
> 
> 
> 
> https://github.com/Expensify/App/assets/20173411/a45e667f-0a7d-4a64-94f4-5a7f15ce8204

Confirmed it's also happening on Staging

eh2077 commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+        // For the first update, we need to check that the previousUpdateID of the fetched update is the same as the lastUpdateIDAppliedToClient.
+        // For any other updates, we need to check if the previousUpdateID of the current update is found in the deferred updates.
+        // If an update is chained, we can add it to the applicable updates.
+        const isChained = isFirst ? update.previousUpdateID === lastUpdateIDAppliedToClient : updates[Number(update.previousUpdateID)];
```
```suggestion
        const isChained = isFirst ? update.previousUpdateID === lastUpdateIDAppliedToClient : !!updates[Number(update.previousUpdateID)];
```

eh2077 commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+// we need to check if there are any gaps between deferred updates.
+type DetectGapAndSplitResult = {applicableUpdates: DeferredUpdatesDictionary; updatesAfterGaps: DeferredUpdatesDictionary; latestMissingUpdateID: number | undefined};
+function detectGapsAndSplit(updates: DeferredUpdatesDictionary): DetectGapAndSplitResult {
+    const updateValues = Object.values(updates);
```
Do we need to sort updateValues by `previousUpdateID` ? Or they're sorted already?

danieldoglas replied:
They are sorted by previousUpdateID. When you create an object, it is ordered by the keys - and then when you use `Object.values` it will create an ordered array based on those keys.

eh2077 commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+
+// This function will check for gaps in the deferred updates and
+// apply the updates in order after the missing updates are fetched and applied
+function validateAndApplyDeferredUpdates(): Promise<Response[] | void> {
```
I think a few examples or test cases here will very helpful to understand the code.

eh2077 commented on src/libs/actions/OnyxUpdateManager.ts:
```diff
@@ -27,6 +28,134 @@ Onyx.connect({
+// In order for the deferred updates to be applied correctly in order,
+// we need to check if there are any gaps between deferred updates.
+type DetectGapAndSplitResult = {applicableUpdates: DeferredUpdatesDictionary; updatesAfterGaps: DeferredUpdatesDictionary; latestMissingUpdateID: number | undefined};
+function detectGapsAndSplit(updates: DeferredUpdatesDictionary): DetectGapAndSplitResult {
```
I think a few examples or test cases here will very helpful to understand the code.

hungvu193 commented:
Ok, I can confirm the logic worked on test branch

https://github.com/Expensify/App/assets/16502320/b4ba98ed-6696-4116-9141-3d1e91d7ef83

danieldoglas commented:
@hungvu193 @eh2077 please also confirm through the network tab that we're not executing several GetMissingOnyxUpdates at the same time while you're testing this

hungvu193 commented:
> @hungvu193 @eh2077 please also confirm through the network tab that we're not executing several GetMissingOnyxUpdates at the same time while you're testing this

Yep. I can confirm it as well.

https://github.com/Expensify/App/assets/16502320/d36eb9a3-2128-4785-ac96-bc3721205450

chrispader committed:
Update OnyxUpdateManager.ts

Co-authored-by: Eric Han <117511920+eh2077@users.noreply.github.com>

Commit:
Merge branch 'main' into @chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages

danieldoglas commented:
@eh2077 , anything else you've noted on this PR that we might need to change?

eh2077 commented:
@danieldoglas Sorry for the delay. I'm on it and let me look into the code and do some more tests. Will update in an hour.

eh2077 commented:
@chrispader I tested the case described in primary GH https://github.com/Expensify/App/issues/38748. I found that sometimes GetMissingOnyxMessages was called two times.

https://github.com/Expensify/App/assets/117511920/c5cbdd62-d225-46ca-8f79-3fbc0ba75aa9

<img width="771" alt="image" src="https://github.com/Expensify/App/assets/117511920/9be1f699-616c-44e2-b4e4-ed0628a43bb9">
<img width="771" alt="image" src="https://github.com/Expensify/App/assets/117511920/6b7f8b29-b14e-4a15-b714-84af9eefcb7e">

chrispader commented:
> @chrispader I tested the case described in primary GH #38748. I found that sometimes GetMissingOnyxMessages was called two times.
>
> And the console log FYI

This is intentional, when we find a gap in the deferred updates. We will then trigger a second call to `GetMissingOnyxMessages`, and potentially even more, until all gaps have been resolved. This is a recursive call in https://github.com/margelo/expensify-app-fork/blob/cd48514a02629f49f39d769425e6b866a8d9a4b6/src/libs/actions/OnyxUpdateManager.ts#L150

eh2077 commented:
@chrispader I see. So, multiple API calls are expected as long as there's no overlapping of the updateID ranges of them, right?

chrispader commented:
> @chrispader I see. So, multiple API calls are expected as long as there's no overlapping of the updateID ranges of them, right?

yes, basically. It might potentially be the case though, that the `updateIDTo` of the first call is the `updateIDFrom` of the second call, in case the first deferred update already creates a gap.

eh2077 commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

eh2077 reviewed (approved):
Looks good and tested well.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

kavimuru commented:
@danieldoglas @hungvu193 Could you explain the step 3 in `Tests 2 `section?

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8598730206) to production by https://github.com/Julesssss in version: 1.4.60-13 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
